xdem/dem.py

```diff
@@ -497,24 +509,105 @@ def coregister_3d(
+
+        # Check inputs
+        if not isinstance(coreg_method, Coreg):
+            raise ValueError("Argument `coreg_method` must be an xdem.coreg instance (e.g. xdem.coreg.NuthKaab()).")
+        if grid not in ["ref", "src"]:
+            raise ValueError(f"Argument `grid` must be either 'ref' or 'src' - currently set to {grid}.")
+
+        # # Ensure that if an initial shift is provided, at least one coregistration method is affine.
+        if estimated_initial_shift:
+            if not (
+                isinstance(estimated_initial_shift, (list, tuple))
+                and len(estimated_initial_shift) == 2
+                and all(isinstance(val, (float, int)) for val in estimated_initial_shift)
+            ):
+                raise ValueError(
+                    "Argument `estimated_initial_shift` must be a list or tuple of exactly two numerical values."
+                )
+            if isinstance(coreg_method, CoregPipeline):
+                if not any(isinstance(step, AffineCoreg) for step in coreg_method.pipeline):
+                    raise TypeError(
+                        "An initial shift has been provided, but none of the coregistration methods in the pipeline "
+                        "are affine. At least one affine coregistration method (e.g., AffineCoreg) is required."
+                    )
+            elif not isinstance(coreg_method, AffineCoreg):
+                raise TypeError(
+                    "An initial shift has been provided, but the coregistration method is not affine. "
+                    "An affine coregistration method (e.g., AffineCoreg) is required."
+                )
+
+            # convert shift
+            shift_x = estimated_initial_shift[0] * reference_elev.res[0]
+            shift_y = estimated_initial_shift[1] * reference_elev.res[1]
+
+            # Apply the shift to the source dem
+            reference_elev = reference_elev.translate(shift_x, shift_y)
+
+        if grid == "ref":
+            src_dem = src_dem.reproject(reference_elev, silent=True)
+        elif grid == "src":
+            reference_elev = reference_elev.reproject(src_dem, silent=True)
 
         coreg_method.fit(
-            reference_elev=reference_elev,
-            to_be_aligned_elev=self,
-            inlier_mask=inlier_mask,
+            reference_elev,
+            src_dem,
+            inlier_mask,
+            random_state=random_state,
             bias_vars=bias_vars,
             **kwargs,
         )
-        return coreg_method.apply(self)  # type: ignore
+
+        aligned_dem = coreg_method.apply(src_dem, resample=resample, resampling=resampling)
+
+        if aligned_reprojected:
+            # Reprojection needed for statistics and plots
+            aligned_dem = aligned_dem.reproject(reference_elev, silent=True)
+
+        # # Add the initial shift to the calculated shift
+        if estimated_initial_shift:
+
+            def update_shift(
```

Comment:
why is it needed to have a function for this? It could be directly in the code without a function.